contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,1242 @@
+    mapping(uint256 => uint256) public weightForBucket;
+
+    /**
+     * The critical bucket is:
+     * - Greater-than-or-equal-to The highest bucket with Outstanding Principal
+     * - Less-than-or-equal-to the lowest bucket with Available Amount
+     *
+     * It is equal to both of these values in most cases except in an edge cases where the two
+     * buckets are different. This value is cached to find such a bucket faster than looping through
+     * all possible buckets.
+     */
+    uint256 public criticalBucket = 0;
+
+    /**
+     * Latest cached value for totalOwedTokenRepaidToLender.
+     * This number updates on the dYdX Margin base protocol whenever the position is
+     * partially-closed, but this contract is not notified at that time. Therefore, it is updated
+     * upon increasing the position or when depositing/withdrawing
+     */
+    uint256 public cachedRepaidAmount = 0;
+
+    // True if the position was closed from force-recovering the collateral
+    bool public wasForceClosed = false;
+
+    // ============ Constants ============
+
+    // Unique ID of the position
+    bytes32 public POSITION_ID;
+
+    // Address of the token held in the position as collateral
+    address public HELD_TOKEN;
+
+    // Address of the token being lent
+    address public OWED_TOKEN;
+
+    // Time between new buckets
+    uint32 public BUCKET_TIME;
+
+    // Interest rate of the position
+    uint32 public INTEREST_RATE;
+
+    // Interest period of the position
+    uint32 public INTEREST_PERIOD;
+
+    // Maximum duration of the position
+    uint32 public MAX_DURATION;
+
+    // Margin-call time-limit of the position
+    uint32 public CALL_TIMELIMIT;
+
+    // (NUMERATOR/DENOMINATOR) denotes the minimum collateralization ratio of the position
+    uint32 public MIN_HELD_TOKEN_NUMERATOR;
+    uint32 public MIN_HELD_TOKEN_DENOMINATOR;
+
+    // Accounts that are permitted to margin-call positions (or cancel the margin call)
+    mapping(address => bool) public TRUSTED_MARGIN_CALLERS;
+
+    // ============ Constructor ============
+
+    constructor(
+        address margin,
+        bytes32 positionId,
+        address heldToken,
+        address owedToken,
+        uint32[7] parameters,
+        address[] trustedMarginCallers
+    )
+        public
+        OnlyMargin(margin)
+    {
+        POSITION_ID = positionId;
+        HELD_TOKEN = heldToken;
+        OWED_TOKEN = owedToken;
+
+        BUCKET_TIME = parameters[0];
+        INTEREST_RATE = parameters[1];
+        INTEREST_PERIOD = parameters[2];
+        MAX_DURATION = parameters[3];
+        CALL_TIMELIMIT = parameters[4];
+        MIN_HELD_TOKEN_NUMERATOR = parameters[5];
+        MIN_HELD_TOKEN_DENOMINATOR = parameters[6];
+
+        for (uint256 i = 0; i < trustedMarginCallers.length; i++) {
+            TRUSTED_MARGIN_CALLERS[trustedMarginCallers[i]] = true;
+        }
+
+        // Set maximum allowance on proxy
+        OWED_TOKEN.approve(
```

Comment:
Forget if I've asked this: what happens if we eat through all this allowance? Do we refresh it anywhere?